### DISCUSSION TITLE
Stop CLP after a given amount of time

### DISCUSSION THREAD

**Discussion by jhmgoossens** [coin-or/Clp#186] [Q&A] [answered]
I would like to see a way to stop CLP after a given amount of time (clock time or CPU time) has elapsed, and return with the best solution so far and a status indicating that CLP ran out of time. We use Clp in a production environment, with a periodic calibration run that does thousands of similar but independent LPs, usually scheduled for a weekend and taking all available cores (~ 100). Most of the LPs solve in under 2 minutes but several take over half an hour of computation to return a solution. Today I killed one LP that was still running after using over 10 hours of CPU time without returning from the call to CLP Simplex solver. Solving this particular LP with the Barrier method returned a solution in under 5 minutes. Currently the only way I see of implementing this time limit is by imposing an external CPU time limit on the process performing the computation and then detecting that the operating system killed the process when it has used up its time quota. This is not an ideal (portable) solution. Portability is highly desired, since this mission-critical application is over 30 years old and has survived porting across 4 different CPU architectures (currently running on x86-64-bit Linux). It started using CLP as the solver around 8 years ago.

_Originally posted by @ralacroix in https://github.com/coin-or/Clp/discussions/168#discussioncomment-577340_

**Comment by jhmgoossens** [comment, marked as answer]
@ralacroix Is using setMaximumSeconds not an option?
```
ClpSimplex clp;
clp.setMaximumSeconds(timelimit);
```
(Discussion moved from comment on #168)

**Reply by ralacroix** [reply]
I will try this. I will need to find out how to do this through the OSI as this is the interface used exclusively. Right from the start OSI made our application work with CLP and with CPLEX.

**Reply by jhmgoossens** [reply]
Indeed, doing this via Osi you’ll need to split this yourself, since Osi doesn’t include something like setTimeLimit, as discussed https://github.com/coin-or/Osi/discussions/150

**Reply by ralacroix** [reply]
This works great! Thanks.
In the place where I initialize the CLP-specific part of the OSI interface I simply inserted one line of code.
// typedef OsiClpSolverInterface OsiXxxSolverInterface;
// OsiXxxSolverInterface *si = new OsiXxxSolverInterface();
// double time_limit_seconds;
si->getModelPtr()->setMaximumSeconds(time_limit_seconds);
